bookkeeper-server/src/test/java/org/apache/bookkeeper/client/MockBookKeeperTestCase.java

```diff
@@ -0,0 +1,390 @@
+        when(bk.getRecoverOpLogger()).thenReturn(nullStatsLogger.getOpStatsLogger("mock"));
+        when(bk.getAddOpLogger()).thenReturn(nullStatsLogger.getOpStatsLogger("mock"));
+        when(bk.getReadOpLogger()).thenReturn(nullStatsLogger.getOpStatsLogger("mock"));
+        when(bk.getDeleteOpLogger()).thenReturn(nullStatsLogger.getOpStatsLogger("mock"));
+        when(bk.getCreateOpLogger()).thenReturn(nullStatsLogger.getOpStatsLogger("mock"));
+        when(bk.getRecoverAddCountLogger()).thenReturn(nullStatsLogger.getOpStatsLogger("mock"));
+        when(bk.getRecoverReadCountLogger()).thenReturn(nullStatsLogger.getOpStatsLogger("mock"));
+        return nullStatsLogger;
+    }
+
+    @After
+    public void tearDown() {
+        scheduler.shutdown();
+        executor.shutdown();
+    }
+
+    protected void setBookkeeperConfig(ClientConfiguration config) {
+        when(bk.getConf()).thenReturn(config);
+    }
+
+    protected CreateBuilder newCreateLedgerOp() {
+        return new LedgerCreateOp.CreateBuilderImpl(bk);
+    }
+
+    protected OpenBuilder newOpenLedgerOp() {
+        return new LedgerOpenOp.OpenBuilderImpl(bk);
+    }
+
+    protected DeleteBuilder newDeleteLedgerOp() {
+        return new LedgerDeleteOp.DeleteBuilderImpl(bk);
+    }
+
+    protected void closeBookkeeper() {
+        when(bk.isClosed()).thenReturn(true);
+    }
+
+    protected BookieSocketAddress generateBookieSocketAddress(int index) {
+        return new BookieSocketAddress("localhost", 1111 + index);
+    }
+
+    protected ArrayList<BookieSocketAddress> generateNewEnsemble(int ensembleSize) {
+        ArrayList<BookieSocketAddress> ensemble = new ArrayList<>(ensembleSize);
+        for (int i = 0; i < ensembleSize; i++) {
+            ensemble.add(generateBookieSocketAddress(i));
+        }
+        return ensemble;
+    }
+
+    private void setupBookieWatcherForNewEnsemble() throws BKException.BKNotEnoughBookiesException {
+        when(bookieWatcher.newEnsemble(anyInt(), anyInt(), anyInt(), any()))
+            .thenAnswer((Answer<ArrayList<BookieSocketAddress>>) new Answer<ArrayList<BookieSocketAddress>>() {
+                @Override
+                @SuppressWarnings("unchecked")
+                public ArrayList<BookieSocketAddress> answer(InvocationOnMock invocation) throws Throwable {
+                    Object[] args = invocation.getArguments();
+                    int ensembleSize = (Integer) args[0];
+                    return generateNewEnsemble(ensembleSize);
+                }
+            });
+    }
+
+    protected void setupBookieClientAddEntry() {
+        doAnswer((Answer) (InvocationOnMock invokation) -> {
+            Object[] args = invokation.getArguments();
+            BookkeeperInternalCallbacks.WriteCallback callback = (BookkeeperInternalCallbacks.WriteCallback) args[5];
+            BookieSocketAddress bookieSocketAddress = (BookieSocketAddress) args[0];
+            long _ledgerId = (Long) args[1];
```

Comment:
I don't think we use "_xyz" for variables. I would stick to be just "ledgerId" to be consistent. I see this pattern has been used in multiple places in this class.